File: src/srcom/worldrecords.c
/*
 * This program gets the number of WRs that a given player (argv[1]) has set. Optionally, a game
 * (argv[2]) can be specified.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "srcom/utils.h"
#include "srcom/worldrecords.h"

static void
usage(void)
{
	fputs("Usage: `+worldrecords [PLAYER NAME] [GAME (OPTIONAL)]`\n"
	      "Example: `+worldrecords AnInternetTroll mkw`\n",
	      stderr);
	exit(EXIT_FAILURE);
}

int
main(int argc, char **argv)
{
	if (argc == 1)
		usage();

	char *uid = get_uid(argv[1]);
	if (!uid) {
		fprintf(stderr, "Error: User with username '%s' not found.\n", argv[1]);
		exit(EXIT_FAILURE);
	}

	char uri[URIBUF];
	struct game_t *game = NULL;
	if (argc > 2 && (game = get_game(argv[2])) == NULL) {
		fprintf(stderr,
			"Error: Game with abbreviation '%s' not found.\n",
			argv[2]);
		exit(EXIT_FAILURE);
	}

	string_t runs;
	init_string(&runs);

	/* Get players PRs. */
	sprintf(uri, API "/users/%s/personal-bests?top=1&game=%s", uid,
	         game ? game->id : "");
	get_req(uri, &runs);

	/*
	 * Each run will have a level. If the level is null, the run is a
	 * fullgame run. Counting the number of "levels" is a quick and easy way
	 * to get the number of runs without the need of parsing the JSON.
	 */
	const unsigned int total = count_substr(runs.ptr, TOTAL_KEY, TOTAL_KEY_LEN);
	const unsigned int fullgame = count_substr(runs.ptr, LEVEL_KEY, LEVEL_KEY_LEN);

	if (argc >= 3)
		printf("World Record Count: %s - %s\n", argv[1], game->name);
	else
		printf("World Record Count: %s\n", argv[1]);

	printf("Full Game: %u\n"
	       "Individual Level: %u\n"
	       "Total: %u\n",
	       fullgame, total - fullgame, total);

#ifdef DEBUG
	free(runs.ptr);
#endif
	return EXIT_SUCCESS;
}

File: src/srcom/worldrecords.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
)

type JsonData struct {
	Data []struct {
		Run struct {
			Level string `json:"level"`
		} `json:"run"`
	} `json:"data"`
}

func get_counts(json_bytes []byte) (uint, uint) {
	var data JsonData
	err := json.Unmarshal(json_bytes, &data)
	if err != nil {
		ErrorAndDie(err)
	}

	var main uint
	var il uint
	for _, run := range data.Data {
		if run.Run.Level == "" {
			main++
		} else {
			il++
		}
	}

	return main, il
}

func main() {
	argc := len(os.Args)
	if argc == 1 || argc > 3 {
		Usage("worldrecords", "[PLAYER] [GAME (Optional)]", "ElTreago mcbe")
	}

	uid, err := UserID(os.Args[1])
	if err != nil {
		ErrorAndDie(err)
	}
	if argc == 3 {
		game, err := GameData(os.Args[2])
		if err != nil {
			ErrorAndDie(err)
		}
		json_bytes := Request("/users/" + uid + "/personal-bests?top=1&game=" + game.id)
		main, il := get_counts(json_bytes)
		fmt.Printf("World Record Count: %s - %s\n" +
			   "Full Game: %d\n" +
			   "Individual Level: %d\n" +
			   "Total: %d\n", os.Args[1], game.name, main, il, main + il)
	} else {
		json_bytes := Request("/users/" + uid + "/personal-bests?top=1")
		main, il := get_counts(json_bytes)
		fmt.Printf("World Record Count: %s\n" +
			   "Full Game: %d\n" +
			   "Individual Level: %d\n" +
			   "Total: %d\n", os.Args[1], main, il, main + il)
	}
}
